### DISCUSSION TITLE
Negative Jacobian During Contact Problem

### DISCUSSION THREAD

**Discussion by Vandenbg** [idaholab/moose#22967] [Q&A General] [answered]
I've been plagued with this issue for some time now in one form or another. I am modeling nanoindentation and I will occasionally see a negative Jacobian error as my tip is pressed into the material. I have tuned my contact parameters to a point where there is nearly 0 visible distortion under the tip, however, I am still getting this negative Jacobian in some simulations. Here is a picture of my material surface:
![image](https://user-images.githubusercontent.com/91214136/207739933-a018c4f2-39d2-4be5-b0c0-fc5ad43a7cb7.png)

Within my material the mesh is updated each time step. The elements MOOSE creates when updating the mesh are sometimes very small splinters. In the picture below, I focus on a region which has the element that caused the error.
![image](https://user-images.githubusercontent.com/91214136/207741236-6b8ee33a-42d8-4c6b-927e-e9ea434df65b.png)
Is there any parameters that I can use to control when and how MOOSE updates the mesh elements? Should I continue to tune my contact module parameters to a point where this is 0 visual surface deformation? I have run this simulation many times and this problem only arises with some meshes, even though they have all been prepared very similarly.  Any advice would be greatly appreciated. 

Possibly relevant portions of my input deck:
```
[Mesh]
  file = berk_3umpuck_rev2.e
  patch_update_strategy = auto
  #patch_size = 20
[]

[Functions]
  [push_down]
    type = ParsedFunction
    value = 'if(t < 2, -t/4000, -0.0005)'
  []
[]

[Modules/TensorMechanics/Master]
  [arbitrary_name]
    add_variables = true
    strain = FINITE
    block = '1 2'
    use_automatic_differentiation = false
    generate_output = 'stress_xx stress_xy stress_xz stress_yy stress_zz'
    save_in = 'saved_x saved_y saved_z'
    use_finite_deform_jacobian = true
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_type'
  petsc_options_value = 'lu    superlu_dist'
  line_search = 'none'
  l_max_its = 80
  nl_max_its = 30
  dt = 0.004
  dtmin = 0.00005
  end_time = 8
  nl_rel_tol = 1e-6
  nl_abs_tol = 1e-6
  l_tol = 1e-4
  automatic_scaling = true
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Dampers]
  [contact_slip]
    type = ContactSlipDamper
    primary = '202'
    secondary = '102'
  []
[]

[Contact]
  [ind_base]
    primary = '202'
    secondary = '102'
    model = coulomb
    friction_coefficient = 0.4
    normalize_penalty = true
    formulation = tangential_penalty
    penalty = 1e8
    tangential_tolerance = 1
  []
[]
```

**Comment by GiudGiud** [comment]
@recuero @bwspenc @jbtompkins @hugary1995 on advice for meshes undergoing contact deformation
Do we have a doc page on this?

**Comment by hugary1995** [comment]
> Within my material the mesh is updated each time step.

That sounds horrifying... I hope you meant something else.

My usual suggestion to this sort of problem is: use AD or our new system which gives you the exact Jacobian, and switch from pjfnk to Newton.

From there on, if the issue persists, other contact experts will give you more specific suggestions. The best setting, though, is usually problem dependent.

**Reply by recuero** [reply]
In addition to that,  I'd try using a predictor, which tends to help in some contact problems. 

The problem you show appears to be similar to an example in the contact module. The geometry has sharp corners which may contribute to the inversion of elements during the solves. Resolving the geometry problem with some smooth approximation can help. Not sure what enforcement you are using, but if the negative Jacobians become a barrier to get the simulation to complete, I'd try penalty or relaxing the penalty factor.

**Reply by sapitts** [reply]
Gary, thanks for the response. Gavin is working with a crystal plasticity problem that isn't available in AD. Do you know if the ComputeLagrangianWrappedStress will work to call stress update materials outside of the ComputeMultiPlasticityStress class? The ComputeMultipleCrystalPlasticityStress is similar but obviously not the 

Antonio, would you point us to an example of a good predictor to use for this sort of contact problem? Gavin set up his problem following your indenter_berkovich_friction.i example, which uses the SMP preconditioner

**Reply by Vandenbg** [reply]
> In addition to that, I'd try using a predictor, which tends to help in some contact problems.

I can try to set up a predictor.

> The problem you show appears to be similar to an example in the contact module. The geometry has sharp corners which may contribute to the inversion of elements during the solves. 

Originally the example had sharp edges, however, the mesh I created for nanoindentation has a 20nm fillet which largely took care of the negative Jacobian issue. 

> Not sure what enforcement you are using, but if the negative Jacobians become a barrier to get the simulation to complete, I'd try penalty or relaxing the penalty factor.

Thank you for the advice.

**Reply by Vandenbg** [reply]
> > Within my material the mesh is updated each time step.

> That sounds horrifying... I hope you meant something else.

It is horrifying, I am not sure how to interpret this behavior otherwise though. 
t = 0
![image](https://user-images.githubusercontent.com/91214136/207946384-ade7474e-bf0f-4f51-ae22-af1d45701f29.png)

t = 0.08, step 20
![image](https://user-images.githubusercontent.com/91214136/207946633-dbf5ee6d-54bd-4bdc-bf60-2c6d330802b0.png)

t = 0.34, step 85
![image](https://user-images.githubusercontent.com/91214136/207946914-25a47b23-f330-444e-ad4c-341d72135e7a.png)

Does this behavior have a name? or is it something that I shouldn't worry about?

> and switch from pjfnk to Newton.

I can try this.

**Reply by jiangwen84** [reply]
For indentation problem, the deformation is very large which might cause negative Jacobian. You might want to use the elements which are taller in the vertical direction to avoid negative Jacobian. For commercial codes, people typically need to do remeshing when the elements are distorted.

**Reply by bwspenc** [reply]
The mesh is being displaced, and the displaced mesh is updated each iteration. However, we're never modifying the mesh topology. I think what you're seeing in those images is just a Paraview visualization artifact.

**Comment by friedmud** [comment]
How is the boundary condition applied?  Is it a Dirichlet condition on the indenter?  If so - you may need to slow it down.

Generally: I would think that the material would move away from the indentation enough to keep the elements ok... but it looks like you have a decent amount of plasticity or something going on there.

Remeshing (during the solve) is the ultimate answer... but is VERY difficult to do for solid mechanics do to needing to remap the state variables (stateful properties) like accumulated plastic strain, etc.

**Reply by bwspenc** [reply]
I agree with this assessment. I would imagine that if you load this very carefully and use all the best formulation and solver options, you can probably get pretty far (which you already have done), but at a certain point, the distortion in those elements right at the tip will be so high that they will be very close to inverting. One thing that hasn't been mentioned yet which might be helpful is the `ElementJacobianDamper`. It limits the change in the Jacobian during a given nonlinear iteration, which could prevent element inversions when the elements get really distorted.

**Reply by Vandenbg** [reply]
Thank you for the advice, I will look into this damper. I have lowered the time step by half while keeping the total displacement rate the same to see if this helps.

**Reply by hugary1995** [reply]
I'm also curious to know what happened when you encountered negative Jacobian. Did it abort abruptly? Or did it cut the time step and retry?

**Reply by Vandenbg** [reply]
> I'm also curious to know what happened when you encountered negative Jacobian. Did it abort abruptly? Or did it cut the time step and retry?

It would cut the time step every few steps and then eventually get to a spot where even cutting the time step multiple times would cause it to fail.

**Reply by hugary1995** [reply]
Then I guess cutting the time step won't help. Still worth trying though.

**Reply by hugary1995** [reply]
Sorry, to be clear, I'm referring to
> I have lowered the time step by half while keeping the total displacement rate the same to see if this helps.

**Comment by Vandenbg** [comment]
Okay, I have had the opportunity to try several of the suggestions presented here. If you are having issues with the same thing, read all of the replies, you may find something helpful. In my case, the solution ended up being my mesh generation. because I am working at the nano-scale, my mesh is very small. Cubit had a difficult time making a mesh this size that was high quality. The solution was to build the mesh 100x larger and scale after it was complete, this new mesh has less problems than the old ones.

**Reply by GregVernon** [reply, marked as answer]
@Vandenbg - since you mention needing to scale your mesh in Cubit I thought I might add some commentary.  The CAD kernel used by Cubit is ACIS which, like other BREP-based CAD kernels, uses absolute tolerances.  The default absolute tolerance for Cubit is `1E-06` - you can set the accuracy via `set geometry accuracy <double>` where the tolerance is limited to the range: `10.0 <-> 1E-06`.

What we recommend for users is to scale their geometry so that the smallest BREP feature length is ~1.0, though for _most_ "normal engineering" problems (i.e., those that come from their company's CAD/PLM system) it's not strictly necessary.  But for many "physics engineering" problems it is often necessary.  In these cases I usually scale my geometry by the nearest power of 10 for simplicity.  Then after you've built your mesh on the scaled model you can tell the exporter to apply a scale factor _when exporting_ the mesh ([documentation here](https://docs.coreform.com/latest/html/coreform-public/Export.html#%28part._transforming_mesh_coords%29)):

```
transform mesh output scale <factor>
```

Note that `transform` commands are cumulative, added to the effect of previous transforms. To clear a transformation matrix, use the reset option:

```
transform mesh output reset
```

Mesh input and output transformations are also cleared when you reset the entire model using the reset command.

An example script template might look like:

```
reset
## Import geometry and scale
import stp "nanoindenter.stp" noheal
#{scale_factor = 1e3}
volume all scale {scale_factor}
transform mesh output {1/scale_factor}
## Mesh
webcut vol all xplane 
webcut vol all yplane 
imprint all
merge all
mesh vol all
## Export
set exodus netcdf4 off
set large exodus file on
export mesh "nanoindenter.e"  overwrite 
transform mesh output reset
```

**Reply by hugary1995** [reply]
Thanks @GregVernon, very helpful comments!

**Reply by Vandenbg** [reply]
@GregVernon Thank you for the information,

I am curious if the method you outlined is different than simply using `volume all scale #` after meshing to make the mesh smaller. Is it important to scale on the output instead of scaling in my journal file?

**Reply by GregVernon** [reply]
> @GregVernon Thank you for the information,
> 
> I am curious if the method you outlined is different than simply using `volume all scale #` after meshing to make the mesh smaller. Is it important to scale on the output instead of scaling in my journal file?

It is different because `volume all scale <factor>` scales the underlying BREP data, which "owns" the mesh entities.  It's thus possible to corrupt the BREP, which in turn corrupts the mesh attached to it.  It doesn't necessarily always happen, but the difference with using `transform mesh` is that it simply applies the scale factor using double-precision operations applied to the nodal coordinates.  Because it doesn't interact with the BREP, there's **zero** risk of a corrupt BREP messing up your mesh.
